controller/test/frame_detector/frame_detector_test.cpp

```diff
@@ -0,0 +1,198 @@
+#include "frame_detector.h"
+#include "proto_traits.h"
+#include "soft_rx_buffer.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include <string>
+#include <vector>
+using namespace std;
+
+constexpr uint32_t LEN = 20;
+
+using State = FrameDetector<SoftRxBuffer<LEN>, LEN>::State;
+using FrameDetectorTest = FrameDetector<SoftRxBuffer<LEN>, LEN>;
+constexpr uint8_t MARK = static_cast<uint8_t>('.');
+
+SoftRxBuffer<LEN> rx_buf('.');
+
+TEST(FrameDetector, MarkFirstInLost) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+}
+
+TEST(FrameDetector, JunkWhileWaitForStart) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(' ');
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::WAIT_FOR_START_MARKER, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(' ');
+  ASSERT_EQ(State::WAIT_FOR_START_MARKER, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+}
+
+TEST(FrameDetector, RxFullSizeFrame) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  for (uint32_t i = 0; i < LEN - 2; i++) {
+    rx_buf.PutByte(' ');
+  }
+  rx_buf.PutByte(MARK);
+
+  EXPECT_EQ(State::WAIT_FOR_START_MARKER, frame_detector.get_state());
+  ASSERT_TRUE(frame_detector.is_frame_available());
+}
+
+TEST(FrameDetector, RxComplete) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  for (uint32_t i = 0; i < LEN; i++) {
+    rx_buf.PutByte(' ');
+  }
+
+  EXPECT_EQ(State::LOST, frame_detector.get_state());
```

Comment:
Please add a comment why it should be LOST here

Reply:
DONE